src/echoclient/EchoClient.php

```diff
@@ -62,51 +66,60 @@ function __construct()
     }
 
     /**
-     * Add an event to echo
+     * Adds a single event to echo
      *
      * @param string $class
      * @param string $source
      * @param array $props
      * @param string|null $userId
+     * @param string|null $timestamp
      * @return bool True if successful, else false
      */
-    public function createEvent($class, $source, array $props = array(), $userId = null)
+    public function createEvent($class, $source, array $props = array(), $userId = null, $timestamp = null)
     {
-        $class = ECHO_CLASS_PREFIX . $class;
-        $baseUrl = $this->getBaseUrl();
-
-        if (!$baseUrl) {
-            // fail silently when creating events, should not stop user interaction as echo events are collected on a best-endeavours basis
-            $this->getLogger()->warning('Echo server is not defined (missing ECHO_HOST define), not sending event - ' . $class,
-                $props);
+        try {
+            $event = new \echoclient\EchoEvent($class, $source, $props, $userId, $timestamp);
+            return $this->sendBatchEvents([$event]);
+        } catch (\Exception $e) {
             return false;
         }
+    }
 
-        $eventUrl = $baseUrl . '/events';
-        $eventJson = $this->getEventJson($class, $source, $props, $userId);
+    /**
+     * Adds multiple events to echo
+     *
+     * @param EchoEvent[] $events An array of EchoEvent objects
+     * @return bool True if successful
+     * @throws InvalidArgumentException If the event data is malformed in any way
+     * @throws EchoHttpException If the server responded with an error
+     * @throws EchoCouldNotSendException If we were unable to send data to the Echo server
+     */
+    public function sendBatchEvents(array $events)
+    {
+        if (empty($events)) {
+            return true;
+        }
 
-        try {
-            $client = $this->getHttpClient();
-            $request = $client->post($eventUrl, $this->getHeaders(), $eventJson, array('connect_timeout' => 2));
-            $response = $request->send();
+        if (count($events) > self::ECHO_MAX_BATCH_EVENTS) {
+            $this->getLogger()->warning('Batch contains more than ' . self::ECHO_MAX_BATCH_EVENTS . ' events');
+            throw new \InvalidArgumentException("Batch of events exceeds the maximum allowed size");
```

Comment:
If we type these exceptions:

* `TooManyEventsInBatchException extends InvalidArgumentException`
* `PayloadTooLargeException extends InvalidArgumentException`
* `BadEventDataException extends InvalidArgumentException`

the client can programatically choose how to deal with these.  If they're all `InvalidArgumentException`, the client cannot really choose between them (exception messages are not fit for this purpose).

Would it be sensible to introduce exception types to make this clearer?

Reply:
Would make sense to introduce specific exception types.  Hitting the maximum size limit is not an InvalidArgumentException.

Reply:
Why not?  Is there a better exception?